src/messages_manager.ts

```diff
@@ -24,62 +45,106 @@ export class MessagesManager implements Disposable {
     private async handleRedraw({ name, args }: EventBusData<"redraw">): Promise<void> {
         switch (name) {
             case "msg_show": {
-                const hasReturnPrompt = args.some(([type]) => type === "return_prompt");
-                const msg = args.reduce((acc, [type, content, clear]) => {
+                let lineCount = 0;
+
+                for (const [type, content, replaceLast] of args) {
                     // Ignore return_prompt
                     //
                     // A note to future readers: return_prompt is sent much more often with ui_messages. It may
                     // not do what you expect from what :help ui says, so be careful about using these events.
                     // See: https://github.com/vscode-neovim/vscode-neovim/issues/2046#issuecomment-2144175058
-                    if (type === "return_prompt") return acc;
-                    if (clear) return "";
-                    return acc + content.map((c) => c[1]).join("");
-                }, "");
-                const outputMsg = hasReturnPrompt ? msg.replace(/\n$/, "") : msg;
+                    if (type === "return_prompt") {
+                        this.replaceOutput = true;
```

Comment:
I think you might be falling victim to the same bug that I fell victim to (which caused me to write this comment). Reading the issue you linked in the PR description, it seems that the `msg_history_clear` is sent following a `return_prompt`, yeah?

I'd also be curious to know how this affects something like "Pattern not found" messages. Are they still visible at the console at all, or are they immediately cleared?

Reply:
> Reading the issue you linked in the PR description, it seems that the `msg_history_clear` is sent following a `return_prompt`, yeah?

From what I can tell, that's usually true? I guess maybe the clear would normally be sent after the user pressed Enter in response to that message. Actually, that's an interesting find — maybe it would makes sense to have the `return_prompt` "cancel out" a subsequent `msg_clear`. 

I kept iterating on this a little after I pushed the draft, and I found that it worked fairly well to simply skip setting `didChange = true` when a clear comes in — so effectively, we'd still clear the buffer, but avoid redrawing on the next flush, so the user could still see the output in the channel. Maybe limiting that to only happen after a `return_prompt` would be more correct.


> I'd also be curious to know how this affects something like "Pattern not found" messages. Are they still visible at the console at all, or are they immediately cleared?

Yeah, these are followed by a clear command almost immediately as well (with the same `return_prompt`):
```
2024-07-24T13:50:25.500Z MessagesManager: msg_show [
  [ 'wmsg', [ [ 98, 'search hit BOTTOM, continuing at TOP' ] ], true ],
  [
    'emsg',
    [ [ 78, 'E486: Pattern not found: aslkdjfhlasdk' ] ],
    false
  ],
  [
    'return_prompt',
    [ [ 153, 'Press ENTER or type command to continue' ] ],
    false
  ]
]
extensionHostProcess.js:148
2024-07-24T13:50:25.501Z MessagesManager: Flushing message buffer: [
  'search hit BOTTOM, continuing at TOP',
  'E486: Pattern not found: aslkdjfhlasdk'
]
extensionHostProcess.js:148
2024-07-24T13:50:25.502Z MessagesManager: 'search hit BOTTOM, continuing at TOP\nE486: Pattern not found: aslkdjfhlasdk\n'
extensionHostProcess.js:148
2024-07-24T13:50:25.503Z MessagesManager: msg_clear [ [] ]
extensionHostProcess.js:148
2024-07-24T13:50:25.505Z MessagesManager: Flushing message buffer: []
```